apis/eks/v1beta1/types.go

```diff
@@ -0,0 +1,313 @@
+const (
+	ClusterStatusCreating ClusterStatusType = "CREATING"
+	ClusterStatusActive   ClusterStatusType = "ACTIVE"
+	ClusterStatusDeleting ClusterStatusType = "DELETING"
+	ClusterStatusFailed   ClusterStatusType = "FAILED"
+	ClusterStatusUpdating ClusterStatusType = "UPDATING"
+)
+
+// LogType is a type of logging.
+type LogType string
+
+// Log types.
+const (
+	LogTypeAPI               LogType = "api"
+	LogTypeAudit             LogType = "audit"
+	LogTypeAuthenticator     LogType = "authenticator"
+	LogTypeControllerManager LogType = "controllerManager"
+	LogTypeScheduler         LogType = "scheduler"
+)
+
+// ClusterParameters define the desired state of an AWS Elastic Kubernetes
+// Service cluster.
+type ClusterParameters struct {
+	// The encryption configuration for the cluster.
+	// +immutable
+	// +optional
+	EncryptionConfig []EncryptionConfig `json:"encryptionConfig,omitempty"`
+
+	// Enable or disable exporting the Kubernetes control plane logs for your cluster
+	// to CloudWatch Logs. By default, cluster control plane logs aren't exported
+	// to CloudWatch Logs. For more information, see Amazon EKS Cluster Control
+	// Plane Logs (https://docs.aws.amazon.com/eks/latest/userguide/control-plane-logs.html)
+	// in the Amazon EKS User Guide .
+	//
+	// CloudWatch Logs ingestion, archive storage, and data scanning rates apply
+	// to exported control plane logs. For more information, see Amazon CloudWatch
+	// Pricing (http://aws.amazon.com/cloudwatch/pricing/).
+	// +optional
+	Logging *Logging `json:"logging,omitempty"`
+
+	// The VPC configuration used by the cluster control plane. Amazon EKS VPC resources
+	// have specific requirements to work properly with Kubernetes. For more information,
+	// see Cluster VPC Considerations (https://docs.aws.amazon.com/eks/latest/userguide/network_reqs.html)
+	// and Cluster Security Group Considerations (https://docs.aws.amazon.com/eks/latest/userguide/sec-group-reqs.html)
+	// in the Amazon EKS User Guide. You must specify at least two subnets. You
+	// can specify up to five security groups, but we recommend that you use a dedicated
+	// security group for your cluster control plane.
+	//
+	// ResourcesVpcConfig is a required field
+	ResourcesVpcConfig VpcConfigRequest `json:"resourcesVpcConfig"`
+
+	// The Amazon Resource Name (ARN) of the IAM role that provides permissions
+	// for Amazon EKS to make calls to other AWS API operations on your behalf.
+	// For more information, see Amazon EKS Service IAM Role (https://docs.aws.amazon.com/eks/latest/userguide/service_IAM_role.html)
+	// in the Amazon EKS User Guide .
+	//
+	// RoleArn is a required field
+	// +immutable
+	// +optional
+	RoleArn *string `json:"roleArn"`
+
+	// RoleArnRef is a reference to an IAMRole used to set
+	// the RoleArn.
+	// +immutable
+	// +optional
+	RoleArnRef *runtimev1alpha1.Reference `json:"roleArnRef,omitempty"`
+
+	// RoleArnSelector selects references to IAMRole used
+	// to set the RoleArn.
+	// +immutable
```

Comment:
@muvaf agreed, should the reference also be immutable though? What if you reference something that doesn't exist and you want to change it? This actually brings up a larger discussion on these fields. Most of them are not `immutable` after creation of the managed resources, they are `immutable` after creation of the external infrastructure. If you have an invalid parameter for an `immutable` field we should probably let you update it (I realize that right now these immutable tags aren't really doing anything, but in the future they would theoretically represent an immutable field and be checked by API server. Maybe we are okay just saying you have to totally recreate the object if you provide an invalid value for an immutable field?). cc @negz @kasey

Reply:
> agreed, should the reference also be immutable though? What if you reference something that doesn't exist and you want to change it? This actually brings up a larger discussion on these fields.

I think we need to define what's actually immutable here: the actual role that will be attached or vpc of the RDS instance that will be created in. In explicit value and reference cases, user's input is directly that; the actual role whether it's on AWS level (actual value) or Crossplane level (reference). But that's not true for selectors; the input is not the config itself, it's a set of instructions to find one. So, the final value that is set by selector on the reference field should be immutable but there is no much reason for instructions themselves to be immutable. If it finds one then the ref will be populated and that's it, but if it couldn't find one then new labels can be added and the result of that selection should be effective because we'd not be changing the actual config; it was never set.

Reply:
I see your point here, updated to only have the `Selector` not be `// +immutable`.